stan/math/rev/core/precomputed_gradients.hpp

```diff
@@ -17,51 +17,118 @@ namespace math {
  *
  * Stan users should use function precomputed_gradients()
  * directly.
+ *
+ * @tparam ContainerOperands tuple of any container operands (var_value
+ * containing Eigen types)
+ * @tparam ContainerGradients tupleof any container gradients (Eigen types)
  */
-class precomputed_gradients_vari : public vari {
+template <typename ContainerOperands = std::tuple<>,
+          typename ContainerGradients = std::tuple<>>
+class precomputed_gradients_vari_template : public vari {
  protected:
   const size_t size_;
   vari** varis_;
   double* gradients_;
+  static_assert(std::tuple_size<ContainerOperands>::value
+                    == std::tuple_size<ContainerGradients>::value,
+                "precomputed_gradients_vari: ContainerOperands and "
+                "ContainerGradients should have same size!");
+  static constexpr size_t N_containers
+      = std::tuple_size<ContainerOperands>::value;
+  ContainerOperands container_operands_;
+  ContainerGradients container_gradients_;
+
+  /**
+   * Checks that sizes of containers in container operands and
+   * container_gradients match.
+   *
+   * @throws std::invalid_argument sizes do not match
+   */
+  template <size_t... Is>
+  void check_sizes(std::index_sequence<Is...>) {
+    static_cast<void>(std::initializer_list<int>{
+        (check_matching_dims("precomputed_gradients_vari", "operands",
+                             std::get<Is>(container_operands_), "gradients",
+                             std::get<Is>(container_gradients_)),
+         0)...});
+  }
 
  public:
   /**
    * Construct a precomputed vari with the specified value,
-   * operands, and gradients.
+   * operands, gradients and optionally container operands and containers of
+   * gradients.
    *
+   * @tparam ContainerOps tuple of any container operands (var_value
+   * containing Eigen types)
+   * @tparam ContainerGrads tupleof any container gradients (Eigen types)
    * @param[in] val The value of the variable.
    * @param[in] size Size of operands and gradients
    * @param[in] varis Operand implementations.
    * @param[in] gradients Gradients with respect to operands.
+   * @param container_operands any container operands
+   * @param container_gradients any container gradients
    */
-  precomputed_gradients_vari(double val, size_t size, vari** varis,
-                             double* gradients)
-      : vari(val), size_(size), varis_(varis), gradients_(gradients) {}
+  template <typename ContainerOps = std::tuple<>,
+            typename ContainerGrads = std::tuple<>>
+  precomputed_gradients_vari_template(double val, size_t size, vari** varis,
```

Comment:
I think the way to go on this is instead of separating things between `vari**`/`double*` arguments and `std::tuples` in one signature, we do one signature that is only `vari**`/`double` (this one can be for all the precomputed gradient vari stuff we have), and then one that does the tuple thing (this can handle all the operands and partials stuff).

If this thing is named something else and is only used inside operands and partials that's fine.

So the signature here would be:
```cpp
template <typename... T>
precomputed_gradients_vari_template(double val, size_t size,
                    std::tuple<vari_value<T>*...> varis,
                    std::tuple<T*...> gradients) {
  ...
}
```

I guess this is unrelated but related, but all the logic we built up for the variadic stuff is gonna have to graduate to tuples as well for this ([save_varis](https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/save_varis.hpp), [accumulate_adjoints](https://github.com/stan-dev/math/blob/develop/stan/math/rev/core/accumulate_adjoints.hpp) in particular). I don't think we'll want to do a mixed `vari**`/`std::tuple<vari_value<T>*...>` solutions if we can just go all `std::tuple<vari_value<T>*...>`.

Reply:
What you are suggesting would prevent us from using `var_value<Matrix>` and `Matrix<var>` in the same distribution (same `operands_and_partials` instance. 

> . I don't think we'll want to do a mixed vari**/std::tuple<vari_value<T>*...> solutions if we can just go all std::tuple<vari_value<T>*...>

The issue is we can't in general. If we have Eigen or std vector of `var`s, its length is dynamic, so we can't convert it into a tuple (which must have static size).

Reply:
Right right, my bad, but we do this in one of the adj_jac_apply variants and it works now (it's [here](https://github.com/stan-dev/math/blob/f1f32af911b15e3ca4f991b85f03caba9d93caf5/stan/math/rev/functor/adj_jac_apply.hpp#L133)).

So input args of like `var, Eigen<var>, var_value<Eigen>` would make a tuple like `vari*, vari**, vari_value<Eigen>*` or something. So we're still using the `vari**`s, they're just buried in the tuple so it's all the same code. And then everything is packed into the the tuples which is nice.

I guess we can't make operands_and_partials variadic without a bunch of other changes because the code that uses this all makes `.edge1_`, `.edge2_`, etc. accesses.

Reply:
That would be doable. I'm not sure we really need to do it. It gets us a bit simpler interface, but probably a bit less simple logic to handle it. In some extreme cases it might even be a bit slower imo.

I was also thinking about redesingning operands_and_partials from scratch, but it is used in more places than I would like to change.

Reply:
> I was also thinking about redesingning operands_and_partials from scratch, but it is used in more places than I would like to change.

I've been in this same boat. I've thought about semi-meaningful API changes that would be nice but it would be a good bit of work and without performance speedups I found it hard to justify

Reply:
Jeez this was 3 days ago apologies for the delay.

Alright I see the argument let me go back through things again.